migrations/versions/d3515ecea53d_backfill_all_commits.py

```diff
@@ -0,0 +1,115 @@
+"""Backfill all commits
+
+Ensures that all default-branch commits are populated correctly in the database, and
+that the database doesn't have non-default-branch commits that lie and say they're on
+the default branch.
+
+Revision ID: d3515ecea53d
+Revises: 480dbbd48927
+Create Date: 2022-11-09 11:33:51.081418
+
+"""
+import datetime
+import uuid
+from typing import List, Tuple
+
+from alembic import op
+from sqlalchemy import MetaData, Table, distinct, select
+
+from conbench.entities.commit import Commit, GitHub, repository_to_name
+
+# revision identifiers, used by Alembic.
+revision = "d3515ecea53d"
+down_revision = "480dbbd48927"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    connection = op.get_bind()
+    meta = MetaData()
+    meta.reflect(bind=connection)
+    commit_table: Table = meta.tables["commit"]
+    github = GitHub()
+
+    repos: List[Tuple[str]] = list(
+        connection.execute(select(distinct(commit_table.c.repository)))
+    )
+    print(f"All repos: {repos}")
+
+    for (repo,) in repos:
+        if repo != repo.lower():
+            print(f"{repo} has a (deprecated) uppercase letter; skipping")
+            continue
+
+        name = repository_to_name(repo)
+        default_branch = github.get_default_branch(name)
+        all_commits = github.get_commits_to_branch(
+            name=name,
+            branch=default_branch,
+            since=datetime.datetime(1970, 1, 1),
+            until=datetime.datetime.now(),
+        )
```

Comment:
It's interesting that we do potentially quite a bit of HTTP interaction with an external service here in the `upgrade()` function. There are probably a number of assumptions here about the robustness of the interaction, and also about rate limiting / quota errors, in relationship to the number of commits that are actually being fetched. Will this fail hard for repos with more than 5000 commits, or will this enter a long retry loop? (5000 requests per hour per api token is one of the quotas that the github api implements)

Just thinking out loud, :shrug: :) no reply expected.

Reply:
Thanks for writing down these thoughts.

And just maybe for clarification: if one of those ~300 HTTP requests fails unexpectedly (e.g. as of a transient error or as of rate limiting) -- then the migration will probably crash, right? And then some outside system (k8s, for example) will restart the entire thing, and the migration will start again, right?

Reply:
If it fails on the first page for a repo, `get_commits_to_branch()` should return `[]` and this script will remove all the branches from the commits in the database. This is probably okay because we could try the script again and the branches will be re-populated. I'll keep an eye out for this case.

If it fails on a subsequent page, we'll probably get an error like "can't iadd list and NoneType" and the script will crash. The deployment Buildkite job will fail and will have to be manually re-triggered. I'll keep an eye out for this too.

Reply:
> get_commits_to_branch() should return []

Oh. I was hoping this function raises an exception :).